### PR TITLE
Attempt to enable Azure Pipelines as a CI provider

### PR DESCRIPTION
**Pull request by tommyang**
I was following #57156 as I look forward to being able to use Azure Pipelines for Cask taps. 
I see that the activity on that PR has died down recently and thought I would give it a try. Apologies to @shikhar-scs if you are still working on it.

Here are the new changes:
- Updated `vmImage` version. 
- `set -o errexit` and `set -o pipefail` added to the bash script per comments from the previous PR. 
- `TAP_DIR` is set and serves as the rsync destination in the bash script. 
- Use `brew update-reset` on `Homebrew/brew` instead of doing `brew update` per suggestion from the previous PR. (Is this enough?)
- Created an Azure build pipeline _for my forked repo_ to test this. 
  - Note that in order to test with the modified `brewcask-ci.rb` in this PR on Azure, the bash script overwrites `Homebrew/homebrew-cask`'s shipping version. 

Here is where I am stuck:
I noticed that `brewcask-ci.rb` needs commit range info to work. On Travis, this info is provided by the `TRAVIS_COMMIT_RANGE` env variable. Since there is no Azure Pipelines equivalent, commit range info needs to be derived. 
I tried to copy [how homebrew-test-bot does it](https://github.com/Homebrew/homebrew-test-bot/blob/5cf61bcb03bb70ca657724a8d8db6571e5a25abb/cmd/brew-test-bot.rb#L413-L419) but it doesn't seem to work with my test PR tommyang/homebrew-cask#1. ~~[Azure Pipelines log](https://dev.azure.com/tommyang/homebrew-cask/_build/results?buildId=144) indicates that `SYSTEM_PULLREQUEST_TARGETBRANCH` and `BUILD_SOURCEVERSION` were not set.~~ Env variables are filtered. 

Not sure how far I can go from here as I am familiar with neither Ruby nor Azure Pipelines. But I thought I should share my findings.

### CODE REVIEW

File: cmd/brewcask-ci.rb [hunk 1]
```diff
@@ -22,6 +22,25 @@ def run
           raise CaskError, "This command must be run from inside a tap directory."
         end
 
+        current_commit_hash = system_command!(
+          "git", args: ["rev-parse", "--short", "HEAD"]
+        ).stdout.strip
+        # Use Travis CI Git variables
+        if ENV.key?("TRAVIS_COMMIT_RANGE")
+          @commit_range = @commit_range
```
Comment (reitermarkus) [inline cmd/brewcask-ci.rb]: `@commit_range = @commit_range` does nothing.

File: cmd/brewcask-ci.rb [hunk 2]
```diff
@@ -22,6 +22,25 @@ def run
...
+        # Use Azure Pipeline variables
+        elsif ENV.key?("SYSTEM_PULLREQUEST_TARGETBRANCH") && ENV.key?("BUILD_SOURCEVERSION")
+          start_commit_hash = system_command!(
+            "git", args: ["rev-parse", "--short", ENV["SYSTEM_PULLREQUEST_TARGETBRANCH"]]
+          ).stdout.strip
+          @commit_range = "#{start_commit_hash}...#{current_commit_hash}"
+        else
+          @commit_range = "#{current_commit_hash}...#{current_commit_hash}"
```
Comment (reitermarkus) [inline cmd/brewcask-ci.rb]: Move the `@commit_range =` in front of the `if` above.

```ruby
@commit_range = if
  ...
end
```

**Review by reitermarkus — CHANGES_REQUESTED** [review]

File: azure-pipelines.yml [hunk 3]
```diff
@@ -0,0 +1,25 @@
+jobs:
+- job: macOS
+  pool:
+    vmImage: macOS-10.13
+  steps:
+    - bash: |
+        # Force strict error checking.
+        set -o errexit
+        set -o pipefail
```
Comment (reitermarkus) [inline azure-pipelines.yml]: Also add `nounset`.

File: cmd/brewcask-ci.rb [hunk 4]
```diff
@@ -22,6 +22,25 @@ def run
           raise CaskError, "This command must be run from inside a tap directory."
         end
 
+        current_commit_hash = system_command!(
+          "git", args: ["rev-parse", "--short", "HEAD"]
+        ).stdout.strip
```
Comment (reitermarkus) [inline cmd/brewcask-ci.rb]: Add a newline after this.

File: azure-pipelines.yml [hunk 5]
```diff
@@ -12,10 +12,7 @@ jobs:
         export HOMEBREW_COLOR=1
         export HOMEBREW_DEVELOPER=1
         export HOMEBREW_NO_AUTO_UPDATE=1
-        export HOMEBREW_AZURE_REPOSITORY_URI="$(Build.Repository.Uri)"
-        export HOMEBREW_AZURE_PULLREQUESTNUMBER="$(System.PullRequest.PullRequestNumber)"
-        export HOMEBREW_AZURE_TARGETBRANCH="$(System.PullRequest.TargetBranch)"
-        export HOMEBREW_AZURE_SOURCEVERSION="$(Build.SourceVersion)"
+        export HOMEBREW_NO_ENV_FILTERING=1
```
Comment (reitermarkus) [inline azure-pipelines.yml]: This should not be set, please revert to prefixing the variables with `HOMEBREW_`.

File: cmd/brewcask-ci.rb [hunk 6]
```diff
@@ -22,6 +22,43 @@ def run
...
+
+        pr_url =
```
Comment (reitermarkus) [inline cmd/brewcask-ci.rb]: When would this be necessary if we already have `SYSTEM_PULLREQUEST_TARGETBRANCH`?

File: azure-pipelines.yml [hunk 7]
```diff
@@ -0,0 +1,27 @@
...
+        export HOMEBREW_COLOR=1
+        export HOMEBREW_DEVELOPER=1
+        export HOMEBREW_NO_AUTO_UPDATE=1
+        export HOMEBREW_NO_ENV_FILTERING=1
+        # Using brew update-reset and restricting directories is quicker/more robust than brew update
+        brew update-reset "$(brew --repository)"
+        # TEMPORARY: Overwrite brewcask-ci.rb with my version since the changes are not merged yet
+        cp -f "$(Build.SourcesDirectory)/cmd/brewcask-ci.rb" "$(brew --repository)/Library/Taps/homebrew/homebrew-cask/cmd/"
+        # Mirror the repo as a tap.
+        TAP_DIR="$(brew --repository)/Library/Taps/$(Build.Repository.Name)"
```
Comment (reitermarkus) [inline azure-pipelines.yml]: You can call `brew --repository "$(Build.Repository.Name)"` here instead.

File: cmd/brewcask-ci.rb [hunk 8]
```diff
@@ -22,6 +22,43 @@ def run
...
+            "#{current_commit_hash}...#{pr_commit_hash}"
+          elsif ENV.key?("SYSTEM_PULLREQUEST_TARGETBRANCH") && ENV.key?("BUILD_SOURCEVERSION")
+            # Use Azure Pipeline variables for master or branch jobs.
+            start_commit_hash = system_command!(
+              "git", args: ["rev-parse", "--short", ENV["SYSTEM_PULLREQUEST_TARGETBRANCH"]]
+            ).stdout.strip
+            "#{start_commit_hash}...#{current_commit_hash}"
+          else
+            # Otherwise just use the current hash.
+            "#{current_commit_hash}...#{current_commit_hash}"
```
Comment (reitermarkus) [inline cmd/brewcask-ci.rb]: Raise an error here instead.

**Review by reitermarkus — CHANGES_REQUESTED** [review]

**Comment by tommyang** [comment]
I probably won't have time to work on this in the near future. Closing this for now.

**Pull request closed without merging**